pkg/v1/remote/write.go

```diff
@@ -412,8 +412,9 @@ func scopesForUploadingImage(ref name.Reference, layers []v1.Layer) []string {
 
 	for _, l := range layers {
 		if ml, ok := l.(*MountableLayer); ok {
-			// we add push scope for ref.Context() after the loop
-			if ml.Reference.Context() != ref.Context() {
+			// we will add push scope for ref.Context() after the loop.
+			// for now we ask pull scope for references of the same registry
+			if ml.Reference.Context().Registry == ref.Context().Registry {
```

Comment:
I think we want to keep both the old check and the new check, otherwise you could end up with scopes like:

`repository:test-kaniko/test-kaniko:push,pull&repository:test-kaniko/test-kaniko:pull`

... which might confuse a registry.

Reply:
@jonjohnsonjr Ok. I also added a test to validate this behaviour. 

In the mean time, I find (but I am not an expert) the two lines confusing:

```
scopeSet[ml.Reference.Context().Scope(transport.PullScope)] = struct{}{}
...
scopes = append(scopes, ref.Scope(transport.PushScope))
```

Why don't we do

```
scopeSet[ml.Reference.Context().Scope(transport.PullScope)] = struct{}{}
scopes = append(scopes, ref.Context().Scope(transport.PushScope))
...

(or without `Context()` in both instruction?

Reply:
> Why don't we do

Good question!

> (or without `Context()` in both instruction?

I'd prefer this. I'm not sure why we call `Context` on the `MountableLayer` -- I'm guessing the author was thinking "I need read scopes for this layer's _repository_, so I'll call Context", but I believe there's no different. You could just call `ml.Reference.Scope(transport.PullScope)` instead, which I think is cleaner.

Reply:
Ok I've re pushed this change